Release/2.0.0 - Composer 2 Compatibility

JParkinson1991 opened the pull request:
The following PR will make this package compatibile with Composer version 2.

- Added .gitignore
- Track package dependencies via commited composer.locl
- Update Plugin adding missing PluginInterface methods
- Add missing constructor argument to `src/Netresearch/Composer/Patches/Downloader/Composer.php`

A major version bump is required here, due to the changes in `src/Netresearch/Composer/Patches/Downloader/Composer.php`as the package can no longer be compatibile with both composer 1 and 2 simultaneously.

Consumers of this package can still require either version 1 or 2 as no pulic interfaces have changes.

```json
{
    "require": {
        "netresearch/composer-patches-plugin": "^1.0 || ^2.0"
    }
}
```

---

For those wanting to use these changes now following instructions here: https://github.com/netresearch/composer-patches-plugin/pull/36#issuecomment-732773402

JParkinson1991 commented:
Looks like most of the PrettyCI errors are coming from exisiting source, i do not think they should delay the merge of this PR.

JParkinson1991 commented:
For those looking to use these changes prior to this PR being merged do the following:

**Bump the require constraints where this package is used**
```
{
    "require": {
        "netresearch/composer-patches-plugin": "^1.0 || ^2.0"
    },
}
```

**Add the forked repository as a project repository**
_Note: If providing packages via a metapack, add this repository definition to the main project composer.json_

```
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/JParkinson1991/composer-patches-plugin"
        }
    ],
}
```

**Update the package**

```
composer update netresearch/composer-patches-plugin ^2.0
```

CybotTM commented on composer.json:
```diff
@@ -1,5 +1,6 @@
 {
     "name": "netresearch/composer-patches-plugin",
+    "version": "2.0.0",
```
Version is by tagging in the code repository, not by adding it to the composer.json in the code repository.

JParkinson1991 replied:
Removed.

CybotTM commented:
We do not provide a composer.lock file as part of this repository, because this is not a standalone application. Please remove it from your PR.

CybotTM commented:
Hi @JParkinson1991 ,

thanks for your PR - please review my comments.

CybotTM commented:
And please also consider updating the README.

JParkinson1991 commented:
@CybotTM i have removed `composer.lock` and added it to `.gitignore`

Around adding version to `composer.json` I usually add it manually to all plugin's etc i maintain, but that follows my release process which may not be applicable here. It has been removed.

JParkinson1991 commented:
What's blocking the merge/release of this?

CybotTM commented:
Hi @JParkinson1991 ,

> What's blocking the merge/release of this?

Nothing, except the two tiny code formatting issues.

JParkinson1991 commented:
I wonder if it's worth forgetting about PrettyCI for now, given it seems to be abandoned? https://prettyci.com/goodbye

Given the pressing need to have something out there composer 2 compatible do you release now and look to implement code standard enforcement/validation etc (possibly(using github actions) as part of a subsequent release?

hvt commented:
Hey @JParkinson1991 thanks for the work! @CybotTM would a new tag be possible for this work to reach the public :]?
